apps/calendar/js/controllers/notifications.js

```diff
@@ -17,20 +18,33 @@ exports.unobserve = function() {
   messageHandler.responder.off('alarm', exports.onAlarm);
 };
 
-exports.onAlarm = function(alarm) {
+exports.onAlarm = co.wrap(function *(alarm) {
   debug('Will request cpu wake lock...');
   var lock = navigator.requestWakeLock('cpu');
   debug('Received cpu lock. Will issue notification...');
-  return issueNotification(alarm).catch(err => {
-    console.error('controllers/notifications', err.toString());
-  }).then(() => {
+  try {
+    yield issueNotification(alarm);
+  } catch (err) {
+    console.error('[controllers/notifications]', err.toString());
+  } finally {
     // release cpu lock with or without errors
     debug('Will release cpu wake lock...');
     lock.unlock();
-  });
-};
+  }
+
+  var storeFactory = core.storeFactory;
+  var alarmStore = storeFactory.get('Alarm');
+  alarm.fired = true;
+  // Save that we've fired this alarm.
+  yield alarmStore.persist(alarm);
+});
+
```

Comment:
This seems to be the core of the change, the rest is using co and the in the issueNotification, to skip it alarm.fired is set.

I am unfamiliar with the storage model in calendar, but given that, the concern I have is how alarmStore is cleaned up with the alarm stuff. My first impulse would be store if a notification was acted upon by placing that data on the calendar event itself, so that if it is ever purged or deleted, the cleanup happens naturally from there.